api/v1alpha2/awscluster_conversion.go

```diff
@@ -53,6 +53,14 @@ func (src *AWSCluster) ConvertTo(dstRaw conversion.Hub) error { // nolint
 		dst.Spec.SSHKeyName = nil
 	}
 
+	// 	Preserve the pre-v1alpha3 behavior of allowing open access to the Bastion host if this is an upgrade
+	if restored.Spec.Bastion.CIDRBlocks == nil {
+		dst.Spec.Bastion.CIDRBlocks = []string{"0.0.0.0/0"}
```

Comment:
If the restored value is nil, it means the user's intent was no CIDRs. (Restored is the v1alpha3 version.)

Reply:
@ncdc My thought here is that if we upgrade the controller in a v1alpha2 cluster to v1alpha3, the nil value needs to be replaced because it will crash the API server. I don't believe the Default() function will be called. I could be wrong.

And I thought we wanted to use `[]` to mean no rules, and `nil` was for the default. Please correct me if I'm wrong.

Reply:
To add more detail ... and to correct my thought process if I don't understand the conversion process:

We need to support roundtrip between v1alpha3 and v1alpha2. So in this conversion to v1alpha3, IF the annotation is there, it's been roundtripped and we want to restore the CIDRBlocks. If the annotation is not there and the value is `nil` (not 100% sure that is what happens so please correct me if I'm wrong), we are moving from v1alpha2 to v1alpha3 in an upgrade and we want to preserve the v1alpha2 behavior.

That is my thought process above for the if/else.

Reply:
Sorry, multi-tasking bit me again 😄. You are right that `nil` means keep the current 0/0 default.

`restored` is the v1alpha3 copy of the resource stored as json in an annotation. Given that the defaulting logic you're adding in this PR sets it to 0/0 if you pass in `nil`, I don't see it being possible for `restored` to have a `nil` value here.

Reply:
> If the annotation is not there and the value is nil (not 100% sure that is what happens so please correct me if I'm wrong), we are moving from v1alpha2 to v1alpha3 in an upgrade and we want to preserve the v1alpha2 behavior

This is definitely a confusing area 😄. If the annotation is not there, then `ok` on line 47 will be false, and we'll return early on line 48.

Reply:
@ncdc Thanks for the Zoom and clearing up how the conversion works. So I've removed the if/else which is not correct and left the assignment of CIDRBlocks for when we roundtrip and the we want the restored data from the annotation.